CARDS-1728: In PROMs, the print header should display patient name, mrn, dob, and appointment time

marta- opened the pull request:
**Includes**:
* [CARDS-1727](https://phenotips.atlassian.net/browse/CARDS-1727): The printing header should be configurable via extensions
* [CARDS-1728](https://phenotips.atlassian.net/browse/CARDS-1728): In PROMs, the print header should display patient name, mrn, dob, and appointment time

**Testing**:
* Start the **proms** project,
  * create a patient information form and fill in first/last name, dob, mrn
  * create a visit information form for that patient and fill in visit time
  * create some survey forms, 
  * go to **print preview** both in the survey page and in the visit subject page.
  * **The page header of the print preview should display patient name, dob, MRN, and appointment date/time.**

![image](https://user-images.githubusercontent.com/651980/164524482-021b3cae-a8b7-48bd-b299-467d512d6f9b.png)

* After deleting the data (`rm -rf .cards-data`), start **any other project**,
  * create some forms, 
  * go to **print preview** both in the form and subject page.
  * **The page header of the print preview should display the breadcrumbs and creation date** (unchanged from what was being displayed before this PR).

marta- committed:
CARDS-1727: The printing header should be configurable via extensions
Added the `PrintHeader` extension point

marta- committed:
CARDS-1728: In PROMs, the print header should display patient name, mrn, dob, and appointment time

sashaandjic reviewed (approved):
tested - looks good
-----
Proms prints
[Patient CARDS-1728 _ DATA-PRO.pdf](https://github.com/data-team-uhn/cards/files/8526007/Patient.CARDS-1728._.DATA-PRO.pdf)
[CARDS-1728 _ #1_ Physical Health Assessment _ DATA-PRO.pdf](https://github.com/data-team-uhn/cards/files/8526009/CARDS-1728._.1_.Physical.Health.Assessment._.DATA-PRO.pdf)

LFS prints
[CARDS-1728_ Patient information _ LFS Data Core.pdf](https://github.com/data-team-uhn/cards/files/8526013/CARDS-1728_.Patient.information._.LFS.Data.Core.pdf)
[CARDS-1728_ Radiotherapy _ LFS Data Core.pdf](https://github.com/data-team-uhn/cards/files/8526031/CARDS-1728_.Radiotherapy._.LFS.Data.Core.pdf)

k-pace reviewed (approved):
Amazing - this looks great. I'll try and show this morning at their meeting

marta- committed:
CARDS-1728: In PROMs, the print header should display patient name, mrn, dob, and appointment time
CARDS-558: Replace moment.js with luxon (in proms print header only)

marta- commented:
Changes in the last commits:
* rebased on latest dev
* amended commits to use an extension point id format that is consistent with existing core eip ids
* squashed 3 commits related to 1728 for a cleaner commit history
* replaced `moment` with `luxon` for processing dates (only in the newly introduced `PrintHeader` extension for PROMs

There should be no noticeable changes in the UI, except the dates are somewhat more nicely formatted (e.g. `Oct 13, 2022, 1:30 PM` instead of `Oct 13 2022 1:30`)

sashaandjic commented:
re-tested - all good
[Patient 1728-retest _ Visit Visit #1 _ DATA-PRO.pdf](https://github.com/data-team-uhn/cards/files/8535054/Patient.1728-retest._.Visit.Visit.1._.DATA-PRO.pdf)
[1728-retest _ Visit #1_ Visit information _ DATA-PRO.pdf](https://github.com/data-team-uhn/cards/files/8535056/1728-retest._.Visit.1_.Visit.information._.DATA-PRO.pdf)
[1728-retest_ Patient information _ DATA-PRO.pdf](https://github.com/data-team-uhn/cards/files/8535057/1728-retest_.Patient.information._.DATA-PRO.pdf)

sdumitriu commented on proms-resources/clinical-data/pom.xml:
```diff
@@ -48,6 +48,7 @@
               SLING-INF/content/libs/cards/resources/;path:=/libs/cards/resources/;overwrite:=true,
               SLING-INF/content/libs/cards/conf/;path:=/libs/cards/conf/;overwrite:=true,
               SLING-INF/content/apps/cards/config/CopyAnswers/Questionnaires/;path:=/apps/cards/config/CopyAnswers/Questionnaires/;overwrite:=true,
+              SLING-INF/content/apps/cards/config/CopyAnswers/SubjectTypes/;path:=/apps/cards/config/CopyAnswers/SubjectTypes/;overwrite:=true,
```
Every subdirectory should be listed separately, just like `Patient` is on the next line.

marta- replied:
I'm not following. This is so that Patient.json, which is directly in SubjectTypes, would be loaded.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/PrintPreview.jsx:
```diff
@@ -108,7 +116,9 @@ const useStyles = makeStyles(theme => ({
-  const { open, resourcePath, title, breadcrumb, date, subtitle, disablePreview, fullScreen, onClose, ...rest } = props;
+  const { open, resourcePath, resourceData, title, breadcrumb, date, subtitle, disablePreview, fullScreen, onClose, ...rest } = props;
+
+  const [ headerExtensions, setHeaderExtensions ] = useState()
```
```suggestion
  const [ headerExtensions, setHeaderExtensions ] = useState();
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/PrintPreview.jsx:
```diff
@@ -141,7 +151,22 @@ function PrintPreview(props) {
   }, [content]);
 
+  useEffect(() => {
+    getHeaderExtensions()
```
We should add some kind of cache at some point, this is requested many times when opening a patient chart.

marta- replied:
Later.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/PrintPreview.jsx:
```diff
@@ -141,7 +151,22 @@ function PrintPreview(props) {
+  }, [])
+
   let header = (
+    !headerExtensions ? <div className={classes.header}><CircularProgress /></div>
```
Do we really need a circular progress here? It is invisible, since the print preview usually happens much later after the extensions are fetched.

marta- replied:
Feel free to remove it if it seems useless.

sdumitriu reviewed (changes requested):
Regression:
- create patient information + AUDIT forms
- open the patient chart
- click on the print button **next to the Visit sub-subject**
- WSORD

marta- committed:
CARDS-1728 - Bugfix: clicking on a visit's print preview on the patient chart results in white screen

sashaandjic reviewed (approved):
tested

sdumitriu committed:
CARDS-1728: In PROMs, the print header should display patient name, mrn, dob, and appointment time
Move a source JSON in a better place

sdumitriu committed:
[codestyle] All lines should end with semicolon

sdumitriu committed:
CARDS-1728: In PROMs, the print header should display patient name, mrn, dob, and appointment time
Removed circular progress from the print preview

sdumitriu committed:
CARDS-1728: In PROMs, the print header should display patient name, mrn, dob, and appointment time
Fix exception when no subject data is present
